### DISCUSSION TITLE
Loss of isochronous packets, always in pairs.

### DISCUSSION THREAD

**Discussion by BrunoPutzeys** [hathach/tinyusb#2433] [Q&A] [answered]
HS UAC2 project on STM32F723 disco board.
At random intervals, once every few seconds on average, audio packets get lost in pairs. I've put a duo of counters in dcd_cwc2 to track received packets (incremented before d_event_xfer_complete gets called) and sof packets (incremented before dcd_event_sof). So directly in the interrupt handler. During playback the two counters track nicely except the packets received counter falls behind in steps of two. I never lose just one packet, always two. I see no NAKs from the device to the host. 
I'm currently hoping to find enlightenment in ST's wonderful reference manual. I'd be grateful if anyone had any pointers as to what bits to look at to pinpoint the cause. Is my test a guarantee that it's a hardware issue, or is there anything that can go wrong on the software end that could prevent the DOEPINTx XFRC interrupt from going off?

**Comment by BrunoPutzeys** [comment, marked as answer]
Well, this is embarrassing. I've unplugged it from the Beagle and directly into the hub and it hasn't glitched since. Ouch.

**Reply by HiFiPhile** [reply]
Out of topic, but have you tried #2333 and #2328 ?

**Reply by BrunoPutzeys** [reply]
It's really my unfamiliarity with the finesses of git that is getting in the way for me. I wouldn't even know how to fetch a newly revised part of code whilst also retaining any of my own tweaks.

Re 2333, this'll probably horrify everyone, but since I know I'll be struggling for CPU cycles before the project is over (doing SRC and other stuff alongside) I'm now hacking the code to pass any transfers to the audio ep directly to a buffer on the application side, bypassing the stack entirely.
Re 2328, Fifo counting: good to see this coming. I might well want to switch to that because even if lost frames are normally rare, counting MCK cycles does nothing to correct for that. Of course, having done the above, I just made my life hard with this one.

I might still come to my senses. When this hack works I'll verify that the cycle savings is actually worth the hassle.

edit: I have. While streaming the time spent in tud_irq_handler and the time spent in tud_task total 32us (25% load). That's on a 216MHz core, so probably around 10% on a 550MHz one. Guess that with that in mind, hacking into the stack is a rotten way of saving a few cycles. I'll bone up on git so I can follow development better...
